packages/perseus/src/widgets/interactive-graphs/reducer/interactive-graph-reducer.ts

```diff
@@ -507,9 +534,181 @@ function boundAndSnapToGrid(
+    // to ensure that the point is within the bounds of the graph even after snapping to the nearest degree
+    const lower: vec.Vector2 = [
+        range[0][0] + snapStep[0],
+        range[1][0] + snapStep[0],
+    ];
+    const upper: vec.Vector2 = [
+        range[0][1] - snapStep[1],
+        range[1][1] - snapStep[1],
+    ];
+
+    let result = point;
+
+    if (result[0] < lower[0]) {
+        result = [
+            lower[0],
+            result[1] + (lower[0] - result[0]) * Math.tan(angle),
+        ];
+    } else if (result[0] > upper[0]) {
+        result = [
+            upper[0],
+            result[1] - (result[0] - upper[0]) * Math.tan(angle),
+        ];
+    }
+
+    if (result[1] < lower[1]) {
+        result = [
+            result[0] + (lower[1] - result[1]) / Math.tan(angle),
+            lower[1],
+        ];
+    } else if (result[1] > upper[1]) {
+        result = [
+            result[0] - (result[1] - upper[1]) / Math.tan(angle),
+            upper[1],
+        ];
+    }
+
+    return result;
+}
+
+function boundAndSnapAngleEndPoints(
     destinationPoint: vec.Vector2,
-    {range, coords}: {range: [Interval, Interval]; coords: Coord[]},
+    {
+        range,
+        coords,
+        snapDegrees,
+        angleOffsetDeg,
+        snapStep,
+    }: {
+        range: [Interval, Interval];
+        coords: Coord[];
+        snapDegrees?: number;
+        angleOffsetDeg?: number;
+        snapStep: vec.Vector2;
+    },
+    index: number,
+) {
+    const snap = snapDegrees || 1;
+    const offsetDegrees = angleOffsetDeg || 0;
+
+    // Needed to prevent updating the original coords before the checks for
+    // degenerate triangles and overlapping sides
+    const coordsCopy = [...coords];
+
+    // We want to subtract or add the snapStep to the lower and upper bounds
+    // respectively to ensure that the point is within the bounds of the graph
+    // even after snapping to the nearest degree
+    const angleRange = [
+        [range[0][0] + snapStep[0], range[0][1] - snapStep[0]],
+        [range[1][0] + snapStep[1], range[1][1] - snapStep[1]],
+    ] as [Interval, Interval];
+
+    // Takes the destination point and makes sure it is within the bounds of the graph
+    // SnapStep is [0, 0] because we don't want to snap these points to the grid at all
+    const boundPoint = bound({
+        snapStep: [0, 0],
+        range: angleRange,
+        point: destinationPoint,
+    });
+    coordsCopy[index] = boundPoint;
+
+    // Get the vertex of the angle
+    const vertex = coords[1];
+
+    // Gets the angle between the coords and the vertex
+    let angle = GraphUtils.findAngle(coordsCopy[index], vertex);
+
+    // Snap the angle to the nearest multiple of snapDegrees (if provided)
+    angle = Math.round((angle - offsetDegrees) / snap) * snap + offsetDegrees;
+    const distance = GraphUtils.getDistance(coordsCopy[index], vertex);
```

Comment:
Check if there's a better getDistance OR move the logic to the same place as findAngle